1-Draft/RFCNNNN-#Requires-Additions.md

```diff
@@ -0,0 +1,141 @@
+---
+RFC: RFC
+Author: Robert Holt
+Status: Draft
+SupercededBy:
+Version:
+Area: Hash-requires
+Comments Due: 2018-05-15
+Plan to implement: Yes
+---
+
+# \#Requires Additions
+
+Currently, PowerShell's `#requires` statement (or perhaps
+pragma?) supports the following parameters:
+
+* `-Version <N>[.<n>]`, where a minimum PowerShell version can be specified
+* `-PSSnapin <PSSnapin-Name> [-Version <N><n>]`, where a required PowerShell Snapin can be specified
+* `-Modules { <Module-name> | <Hashtable> }`, where PowerShell modules that are required can be specified
+* `-ShellId <ShellId>`, where the required Shell ID can be specified
+* `-RunAsAdministrator`, where the script is required to be run as administrator
+
+These features are documented in [about_Requires](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_requires?view=powershell-6), along with the statement that:
+
+> You can use #Requires statements in any script. You
+> cannot use them in functions, cmdlets, or snap-ins.
+
+Currently however, this is untrue, as `#requires`
+statements are allowed by the parser/tokenizer anywhere in
+a script and then effectively hoisted to the top of that
+script to be checked before any part of the script is
+executed, no matter where they are placed.
+
+This RFC proposes the following changes:
+
+* Emit a warning when parsing scripts where the `#requires`
+  is not at the top, because of the hoisting behavior.
+* Add support for the following new parameters (each
+  independently up for discussion):
+  * `-OS {Windows | Linux | MacOS}`, where an
+    operating system (or possibly combination of them) can
+    be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/3751).
+  * `-MaximumPSVersion <V>[.<v>]`, where a maximum PowerShell
+     version can be specified as required. See [this PowerShell issue](https://github.com/PowerShell/PowerShell/issues/2846).
+  * `-MinimumPSVersion` as an alias of `-MinimumVersion`.
+
+* **Withdrawn**, in favor of `using` statements.
+  ~~`-Assembly <Assembly-name>`, where a .NET assembly can
```

Comment:
Should we remove this?

Reply:
Just kept it for context. Happy to remove if that's better. Or add a `Withdrawn` section to contain these things